8171508: os::jvm_path: error in handling -Dsun.java.launcher.is_altjvm option after 8066474

calvinccheung opened the pull request:
This change fixes two issues in `os::jvm_path()`:

- go back 4 instead of 5 slashes and then checking for "/jre/lib";
- fix the hardcoded "hotspot" sub-directory before the jvm library name.

Passed tiers 1 - 3 testing.

calvinccheung commented:
/label add hotspot-runtime

tstuefe commented:
Hi @calvinccheung, I thought we planned on removing altjvm handling altogether? See mail thread https://mail.openjdk.org/pipermail/hotspot-runtime-dev/2017-April/023205.html

tstuefe commented:
See https://mail.openjdk.org/pipermail/hotspot-runtime-dev/2017-April/023228.html

dholmes-ora commented:
> go back 4 instead of 5 slashes and then checking for "/jre/lib";

But we got rid of the `jre` directory years ago. ??

calvinccheung commented:
> Hi @calvinccheung, I thought we planned on removing altjvm handling altogether? See mail thread https://mail.openjdk.org/pipermail/hotspot-runtime-dev/2017-April/023205.html

I've pushed a commit which removes altjvm handling in JVM. There's small piece of code in the java launcher (java.c) which I didn't remove because:

- there's an internal test which uses altjvm;
- JVM won't start if -XXaltjvm is specified

calvinccheung commented:
> > go back 4 instead of 5 slashes and then checking for "/jre/lib";
> 
> But we got rid of the `jre` directory years ago. ??

I've removed the processing regarding the slashes and the `jre` directory.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -2772,60 +2772,40 @@ void os::jvm_path(char *buf, jint buflen) {
+  // Look for JAVA_HOME in the environment.
+  char* java_home_var = ::getenv("JAVA_HOME");
+  if (java_home_var != nullptr && java_home_var[0] != 0) {
+    char* jrelib_p;
```
Please remove jre from the name - there is no need to mention jre at all here.

calvinccheung replied:
Fixed.
Also removed the `-Dsun.java.launcher.is_altjvm=true` option from `gtestMain.cpp`.

calvinccheung commented:
> This issue should be renamed "Remove -Dsun.java.launcher.is_altjvm option". That requires a CSR request to be filed.
> 
I filed a draft CSR [JDK-8353678](https://bugs.openjdk.org/browse/JDK-8353678).

dholmes-ora commented:
How does the gtest launcher work with these changes? There the libjvm.so is in a completely different place to the JDK. ??

dholmes-ora commented:
Also as discussed offline picking up JAVA_HOME from the environment is not the right thing to do in general. It only worked with -Dsun.java.launcher.is_altjvm because we ensured we set JAVA_HOME when using the property.

calvinccheung commented:
> How does the gtest launcher work with these changes? There the libjvm.so is in a completely different place to the JDK. ??

gtest launcher sets the `-XX:+ExecutingUnitTests` option. My updated fix is to capture the option in `Arguments::process_sun_java_launcher_properties()` so that the `os::jvm_path()` can check if the option is set before using `JAVA_HOME` setting.

dholmes-ora commented:
Can you update the PR description to clarify what is now being done please Calvin.

dholmes-ora reviewed (changes requested):
So basically now we have removed the `-Dsun.java.launcher.is_altjvm` option and replaced its use with gtest by the use of the `executing_unit_tests` functionality. Okay.

A few minor requests. Thanks.

dholmes-ora commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -2772,60 +2772,41 @@ void os::jvm_path(char *buf, jint buflen) {
           return;
         }
-
-        // determine if this is a legacy image or modules image
```
This applies in all versions of the file of course, where it exists.

calvinccheung replied:
Fixed.

calvinccheung commented:
> Can you update the PR description to clarify what is now being done please Calvin.

Updated PR description.

calvinccheung commented on src/hotspot/os/aix/os_aix.cpp:
```diff
@@ -1306,60 +1306,39 @@ void os::jvm_path(char *buf, jint buflen) {
-            return;
-          }
-        }
+    // modules image doesn't have "jre" subdirectory
```
Fixed

calvinccheung commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -2772,60 +2772,41 @@ void os::jvm_path(char *buf, jint buflen) {
-      if (java_home_var != nullptr && java_home_var[0] != 0) {
-        char* jrelib_p;
-        int len;
+  if (Arguments::executing_unit_tests()) {
```
Fixed.

calvinccheung commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -2772,60 +2772,41 @@ void os::jvm_path(char *buf, jint buflen) {
+      if (0 == access(buf, F_OK)) {
+        // Use current module name "libjvm.so"
+        len = (int)strlen(buf);
+        snprintf(buf + len, buflen-len, "/%s/libjvm%s",
```
Fixed.

tstuefe reviewed (commented):
Are the Posix variants all identical? If yes, can we move this to os::posix.cpp? If not, what are the differences?

dholmes-ora commented:
> Are the Posix variants all identical? If yes, can we move this to os::posix.cpp? If not, what are the differences?

AIX has a different dll lookup mechanism, but otherwise these should be pretty much identical. Windows also only really differs in the lookup mechanism. Maybe new RFE to refactor these and share some code?

tstuefe commented:
> > Are the Posix variants all identical? If yes, can we move this to os::posix.cpp? If not, what are the differences?
> 
> AIX has a different dll lookup mechanism, but otherwise these should be pretty much identical. Windows also only really differs in the lookup mechanism. Maybe new RFE to refactor these and share some code?

Okay. The code could also be simpler if it were to use stringStream, spanned over the caller-provided buffer. No character counting and truncation handling needed then (but we would need to add truncation detection to stringStream, but that is really easy).

calvinccheung commented:
> > > Are the Posix variants all identical? If yes, can we move this to os::posix.cpp? If not, what are the differences?
> > 
> > 
> > AIX has a different dll lookup mechanism, but otherwise these should be pretty much identical. Windows also only really differs in the lookup mechanism. Maybe new RFE to refactor these and share some code?
> 
> Okay. The code could also be simpler if it were to use stringStream, spanned over the caller-provided buffer. No character counting and truncation handling needed then (but we would need to add truncation detection to stringStream, but that is really easy).

I agree with David about doing the refactoring in a new RFE. I've simplified the code by using stringStream as you suggested. I didn't add truncation detection to stringStream (which I think could be done in another RFE). However, I added an assert in `os::jvm_path()` to ensure there's no truncation.

tstuefe reviewed (approved):
Nice, this looks good to me. Thank you.

dholmes-ora commented on src/hotspot/os/aix/os_aix.cpp:
```diff
@@ -1319,22 +1320,21 @@ void os::jvm_path(char *buf, jint buflen) {
 
-      // modules image doesn't have "jre" subdirectory
       len = strlen(buf);
       assert(len < buflen, "Ran out of buffer room");
```
```suggestion
      assert(strlen(buf) < buflen, "Ran out of buffer room");
```
and then I think the `len` local is unused.

calvinccheung replied:
Fixed.

dholmes-ora commented on src/hotspot/os/bsd/os_bsd.cpp:
```diff
@@ -1512,34 +1513,30 @@ void os::jvm_path(char *buf, jint buflen) {
       len = strlen(buf);
       assert(len < buflen, "Ran out of buffer space");
-      char* lib_p = buf + len;
+      // Add the appropriate library  and JVM variant subdirs
```
```suggestion
      // Add the appropriate library and JVM variant subdirs
```

calvinccheung replied:
Fixed.

dholmes-ora reviewed (commented):
Looks ago. Some additional possible cleanup now you are using the stringStream.

dholmes-ora reviewed (approved):
Looks good. Thanks

calvinccheung commented:
Thanks @tstuefe, @dholmes-ora for the review.

/integrate
